pkgs/build-support/coq/default.nix

```diff
@@ -0,0 +1,138 @@
+{ lib, stdenv, coqPackages, coq, fetchFromGitHub, fetchFromGitLab, fetchurl }@args:
+let lib = import ./extra-lib.nix {inherit (args) lib;}; in
+with builtins; with lib;
+let
+  isGitHubDomain = d: match "^github.*" d != null;
+  isGitLabDomain = d: match "^gitlab.*" d != null;
+  isGforgeDomain = d: match "^gforge.*" d != null;
+  isPrefixGforgeDomain = d: match ".*\.gforge.*" d != null;
+  meta-fetch = {domain, owner ? "", repo, rev, name ? "source", sha256 ? null, ...}@args:
+    switch domain [
+      { case = isGitHubDomain; out =
+          if args?sha256
+          then fetchFromGitHub {
+            githubBase = domain;
+            inherit owner repo sha256 rev name;
+          }
+          else fetchTarball {
+            url = "https://${domain}/${owner}/${repo}/archive/${rev}.tar.gz";
+            inherit name;
+          };
+      }
+      { case = isGitLabDomain; out =
+          if args?sha256
+          then fetchFromGitLab {inherit domain owner repo sha256 rev name; }
+          else fetchTarball {
+            url = "https://${domain}/${owner}/${repo}/-/archive/${rev}/${repo}-${rev}.tar.gz";
+            inherit name;
+          };
+      }
+      { case = isGforgeDomain; out = let s = lib.splitString "#" rev; in fetchurl {
+            url = "https://${domain}/frs/download.php/file/${lib.head s}/${repo}-${lib.last s}.tar.gz";
+            inherit sha256;
+          };
+      }
+      { case = isPrefixGforgeDomain; out = fetchurl {
+            url = "https://${domain}/releases/${repo}-${rev}.tar.gz";
+            inherit sha256;
+          };
+      }
+    ] null;
+    # coq-clean-src = src: {
```

Comment:
What’s the purpose of this comment?

Reply:
Well spotted, I am still hesitant whether to include an opt in or opt out (can't decide yet) source filtering of files that are known not to be part of coq build process, to stop recompilung everything because a readme or changelog changed... so you have an opinion?

Reply:
Sounds like a useful feature during the development. For packaging, it does not make much sense, does it? So I would say opt-in.

Reply:
indeed, but if one relies on coqPackages to get a "standard" version of a package for a development branch, they would get an unfiltered package, which means their local copy would part from the `package.override {version = "branch"; }`, which is something I tried very hard to enforce (it will prevent repetitions of builds across different CIs), of course, I could always add an extra argument to override (e.g. `filterSource = true;`), but I'm not sure what we have to gain with the opt-in...

Reply:
Actually, I'm not sure what we lose with the opt-out...

Reply:
Then let’s go with opt-out.

Reply:
Actually, I have problems implementing source filtering (no matter opt in or out).
Cf https://discourse.nixos.org/t/filtersource-and-friends-do-not-compose-with-fetchgit-and-friends/10466

Reply:
Given the outcome of the discussion, I think I will not do source filtering yet, maybe in a subsequent PR.